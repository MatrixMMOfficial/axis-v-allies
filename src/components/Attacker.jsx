import React from 'react'
import { UnitFigTableData } from './UnitFigure'
import { air as isAir } from '../selectors/units'

const Attacker = ({ 
  unit, 
  committed, 
  strategicBombing,
  targetIndex, 
  landingSlots,
  mayNotCommitToStrategicBombing,
  mayNotUncommitFromStrategicBombing,
  hasIndustry,
  commitUnits, 
  unCommitUnits, 
  commitToStrategicBombing
}) => {
  const air = isAir(unit)
  strategicBombing = strategicBombing || []
  return (
    <tr>
      <UnitFigTableData unit={unit} />
      <td className="available">
        <CommitButtons
          unit={unit} 
          committed={committed}
          targetIndex={targetIndex}
          air={air}
          landingSlots={landingSlots}
          action={commitUnits} />
        {air && hasIndustry && <AirOptions 
          unit={unit} 
          committed={committed}
          landingSlots={landingSlots}
          targetIndex={targetIndex} 
          action={commitToStrategicBombing}
          mayNotCommitToStrategicBombing={mayNotCommitToStrategicBombing}
        />}
      </td>
      <td className="available">
        <UncommitButtons
          targetIndex={targetIndex}
          committed={committed.filter(id => !strategicBombing.includes(id))}
          action={unCommitUnits} />
        {air && hasIndustry && <UncommitButtons
          unit={unit} 
          committed={strategicBombing.filter(id => unit.ids.includes(id))}
          mayNotUncommitFromStrategicBombing={mayNotUncommitFromStrategicBombing}
          targetIndex={targetIndex} 
          units={strategicBombing}
          action={unCommitUnits} 
        />}
      </td>
    </tr>
  )
}

const CommitButtons = ({ 
  unit, 
  committed,
  targetIndex, 
  action, 
  air, 
  landingSlots,
  mayNotCommitToStrategicBombing
}) => {
  const { distance, ids, originIndex } = unit
  const available = ids.filter(id => !committed.includes(id))
  const qty = available.length
  const allDisabled = qty === 0 || (air && qty > landingSlots) || mayNotCommitToStrategicBombing
  const oneDisabled = qty === 0 || (air && !landingSlots) || mayNotCommitToStrategicBombing
  return (
    <div>
      <input readOnly size={2} value={qty} />
      <button 
        onClick={e => { action(originIndex, targetIndex, available, air && distance)}}
        disabled={allDisabled}
      >&gt;&gt;</button>
      <button 
        onClick={e => { action(originIndex, targetIndex, [available[0]], air && distance)}}
        disabled={oneDisabled}
      >&gt;</button>
    </div>
  )
}

//TODO: prevent carrier from withdrawing if air units are expecting to land on it
const UncommitButtons = ({ 
  unit, 
  committed,
  targetIndex, 
  action,
  mayNotUncommitFromStrategicBombing
}) => {
  const commitQty = committed.length
  return (
    <div>
      <input readOnly size={2} value={commitQty} />
      <button 
        onClick={e => action(targetIndex, committed)}
        disabled={commitQty === 0 || mayNotUncommitFromStrategicBombing} 
      >&lt;&lt;</button>
      <button 
        onClick={e => action(targetIndex, [committed[0]])}
        disabled={commitQty === 0 || mayNotUncommitFromStrategicBombing}
      >&lt;</button>
    </div>
  )
}

const AirOptions = ({ 
  unit, 
  committed,
  targetIndex, 
  landingSlots,
  action,
  mayNotCommitToStrategicBombing
}) => (
  <div>
    <CommitButtons
      unit={unit} 
      air={true}
      committed={committed}
      targetIndex={targetIndex}
      landingSlots={landingSlots}
      action={action}
      mayNotCommitToStrategicBombing={mayNotCommitToStrategicBombing}
    />
    <span>{unit.type.includes('strategic') ? 'Strategic bombing' : 'Bomber escort'}</span>
  </div>
)

export default Attacker
