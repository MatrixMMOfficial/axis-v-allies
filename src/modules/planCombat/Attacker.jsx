import React from 'react';
import Transport from './Transport';
import { UnitFigTableData } from '../../components/UnitFigure';
import { STRATEGIC_BOMB } from '../../actions';
import { unitCount } from '../../lib/unit';

const Attacker = ({ 
  unit, 
  committed, 
  strategicBombing,
  destinationIndex, 
  commitUnits, 
  unCommitUnits, 
  commitAmphibUnits,
  unCommitAmphibUnits,
  viewTransportLoadOptions, 
  landAttack, 
  landingSlots,
  hasIndustry }) => {
  if (unit.name === 'transport') {
    return (
      <Transport 
        unit={unit} 
        committed={committed}
        landAttack={landAttack}
        destinationIndex={destinationIndex}
        commitUnits={commitUnits}
        viewTransportLoadOptions={viewTransportLoadOptions}
        unCommitUnits={unCommitUnits} 
        commitAmphibUnits={commitAmphibUnits}
        unCommitAmphibUnits={unCommitAmphibUnits}
      />
    )
  }
  return (
    <tr>
      <UnitFigTableData unit={unit} />
      <td className="available">
        <CommitButtons
          unit={unit} 
          index={destinationIndex}
          landingSlots={landingSlots}
          action={commitUnits} />
        {unit.air && hasIndustry && <AirOptions 
          unit={unit} 
          hasIndustry={hasIndustry} 
          index={destinationIndex} 
          action={commitUnits}
        />}
      </td>
      <td className="available">
        <UncommitButtons
          unit={unit} 
          index={destinationIndex}
          units={committed}
          action={unCommitUnits} />
        {unit.air && hasIndustry && <UncommitButtons
          unit={unit} 
          index={destinationIndex} 
          units={strategicBombing}
          action={unCommitUnits} 
        />}
      </td>
    </tr>
  )
}

const CommitButtons = ({ unit, index, action, landingSlots, mission }) => {
  const qty = unitCount(unit);
  const allDisabled = qty === 0 || (unit.air && qty > landingSlots)
  const oneDisabled = qty === 0 || (unit.air && !landingSlots)
  return (
    <div>
      <input readOnly size={2} value={qty} />
      <button 
        onClick={e => { action(unit, index, mission, unit.ids)}}
        disabled={allDisabled}
      >&gt;&gt;</button>
      <button 
        onClick={e => { action(unit, index, mission, [unit.ids[0]])}}
        disabled={oneDisabled}
      >&gt;</button>
    </div>
  )
}

const UncommitButtons = ({ unit, index, units, action }) => {
  const commitQty = unitCount(units);
  return (
    <div>
      <input readOnly size={2} value={commitQty} />
      <button 
        onClick={e => action(unit, index, units.ids)}
        disabled={commitQty === 0}
      >&lt;&lt;</button>
      <button 
        onClick={e => action(unit, index, [units.ids[0]])}
        disabled={commitQty === 0}
      >&lt;</button>
    </div>
  )
}

const AirOptions = ({ unit, hasIndustry, index, action }) => {
  let bombOrEscort = hasIndustry;
  if (bombOrEscort) {
    if (unit.name.includes('strategic')) {
      bombOrEscort = <span>Strategic bombing</span>
    } else {
      bombOrEscort = <span>Bomber escort</span>
    }
  }
  return (
    <div>
      {bombOrEscort}
      <CommitButtons
        unit={unit} 
        index={index}
        action={action}
        mission={STRATEGIC_BOMB}/>
    </div>
  )
}

export default Attacker;
