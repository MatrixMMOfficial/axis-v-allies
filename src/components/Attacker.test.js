'use strict'

import React from 'react'
import Attacker from './Attacker'
import renderer from 'react-test-renderer'

it('renders correctly', () => {
  const unit = { type: 'infantry', power: 'Germany', ids: [1] }
  const tree = renderer.create(
    <Attacker unit={unit} committed={[]} />
  ).toJSON()
  expect(tree).toMatchSnapshot()
})

