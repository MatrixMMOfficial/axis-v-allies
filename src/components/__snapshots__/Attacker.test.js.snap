// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly 1`] = `
<tr>
  <td
    className="unit"
  >
    <strong />
    <figure>
      <img
        alt="Germany infantry"
        className={undefined}
        onClick={undefined}
        src="infantry.png"
      />
      <figcaption>
        infantry
         - attacks @ 
        1
      </figcaption>
    </figure>
  </td>
  <td
    className="available"
  >
    <div>
      <input
        readOnly={true}
        size={2}
        value={1}
      />
      <button
        disabled={undefined}
        onClick={[Function]}
      >
        &gt;&gt;
      </button>
      <button
        disabled={undefined}
        onClick={[Function]}
      >
        &gt;
      </button>
    </div>
  </td>
  <td
    className="available"
  >
    <div>
      <input
        readOnly={true}
        size={2}
        value={0}
      />
      <button
        disabled={true}
        onClick={[Function]}
      >
        &lt;&lt;
      </button>
      <button
        disabled={true}
        onClick={[Function]}
      >
        &lt;
      </button>
    </div>
  </td>
</tr>
`;
